# inicio

In [1]:
#solo es para ver si esta tomando mi ambiente
import sys
print(sys.executable)

c:\proye_shada\Comercio\.venv\Scripts\python.exe


In [2]:
# Cargar todas las librerías
import pandas as pd
import os       #para ver si logro llegar al archivo con la ruta de datos

In [3]:
print("Carpeta actual:", os.getcwd())
print("Contenido de la carpeta:", os.listdir())

#C:\proye_shada\Comercio\data\Online_Retail.xlsx
#df_retail = pd.read_excel('data/Online_Retail.xlsx')---------Esta ruta es para python
df_retail = pd.read_excel('../data/Online_Retail.xlsx')

Carpeta actual: c:\proye_shada\Comercio\notebooks
Contenido de la carpeta: ['comercio.ipynb', 'comercio2.ipynb']


In [4]:
# Muestra los nombres de las columnas

print(df_retail.columns)


Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='str')


In [5]:
#se procede a cambiar los nombres de las columnas por unas mas claras
#cambio de nombres de las columnas

nuevos_nombres = {
    'InvoiceNo': 'num_factura',
    'StockCode': 'cod_articulo',
    'Description': 'descripcion',
    'Quantity': 'cantidad',
    'InvoiceDate': 'fh_factura',
    'UnitPrice': 'precio_unitario',
    'CustomerID': 'id_cliente',
    'Country': 'pais'
}

df_retail = df_retail.rename(columns=nuevos_nombres)
print('*' * 120)

print(df_retail.columns.tolist())

************************************************************************************************************************
['num_factura', 'cod_articulo', 'descripcion', 'cantidad', 'fh_factura', 'precio_unitario', 'id_cliente', 'pais']


#### Extracion de solo 5 registros a manera de exploración

In [6]:
print('*' * 120)
print('*' * 40 +'df_retail------- solo los 5 registros' + '*' * 40)
print(df_retail.head())
print('*' * 120)

************************************************************************************************************************
****************************************df_retail------- solo los 5 registros****************************************
  num_factura cod_articulo                          descripcion  cantidad  \
0      536365       85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1      536365        71053                  WHITE METAL LANTERN         6   
2      536365       84406B       CREAM CUPID HEARTS COAT HANGER         8   
3      536365       84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4      536365       84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           fh_factura  precio_unitario  id_cliente            pais  
0 2010-12-01 08:26:00             2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00             3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00             2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00 

#### información de las columnas

In [7]:
print('*' * 120)
print('*' * 40 +'df_retail' + '*' * 40)
#display(df_calls.head())
print('*' * 120)
print(df_retail.info())
print('*' * 120)

************************************************************************************************************************
****************************************df_retail****************************************
************************************************************************************************************************
<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column           Non-Null Count   Dtype         
---  ------           --------------   -----         
 0   num_factura      541909 non-null  object        
 1   cod_articulo     541909 non-null  object        
 2   descripcion      540455 non-null  object        
 3   cantidad         541909 non-null  int64         
 4   fh_factura       541909 non-null  datetime64[us]
 5   precio_unitario  541909 non-null  float64       
 6   id_cliente       406829 non-null  float64       
 7   pais             541909 non-null  str           
dtypes: datetime64[us](1), fl

#### Exploración de los valores ausentes

In [ ]:
#antes de comenzar otra vez con cambios hagamos la copia y trabajemos en ella

df_retail_2 = df_retail.copy()

In [11]:
print('*' * 120)
print('*' * 40 +'df_retail_2------- revisión de valores ausentes' + '*' * 40)
print(df_retail_2.isna().sum())
print('*' * 120)


************************************************************************************************************************
****************************************df_retail_2------- revisión de valores ausentes****************************************
num_factura             0
cod_articulo            0
descripcion          1454
cantidad                0
fh_factura              0
precio_unitario         0
id_cliente         135080
pais                    0
dtype: int64
************************************************************************************************************************


📌 Se observan que hay valores ausentes en la descripción y en el id_cliente, antes de proceder a eliminarlos , vamos a revisar si hay valores duplicados, para realizar una limpieza mas generica creo que lo mas importante es quitar duplicados, por que eso no afectaria mucho que si elimino algunos datos con valores con inconsitencias


Tomemos un punto importante

En análisis de datos, limpiar no significa eliminar todo lo que parece extraño.

Por ejemplo:

* Una cantidad negativa podría ser una devolución.

* Un precio cero podría corresponder a una operación especial.

* Un cliente ausente no significa necesariamente que la venta sea inválida.

* Una descripción ausente sí es un problema de calidad, pero su impacto depende del análisis que quieras realizar.

Revisar si hay valores duplicados¶

In [14]:
#antes que nada vamos a cambiar el tipo flotante de id_cliente
df_retail_2['id_cliente'] = df_retail_2['id_cliente'].astype('Int64')

In [17]:
# Contar duplicados explícitos
print(df_retail_2.duplicated().sum())

5268


In [18]:
df_retail_2 = df_retail_2.drop_duplicates().copy()

In [20]:
df_retail_2[df_retail_2.duplicated(keep=False)].head(10)

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_unitario,id_cliente,pais


📌 Se encuentra que hay 5268 datos duplicados, por lo que procedemos a quitarlos y veamos que pasa esto lo estamos realizando dentro de la copia, no en el dataframe original

In [21]:
#comprobar si hay quedaron valores duplicados
print(df_retail_2.duplicated().sum())

0


vemos si una vez que se quitaron los valores duplicados, como salen los registros

In [ ]:
print('*' * 120)
print('*' * 40 +'df_retail_2------- revisión de valores ausentes' + '*' * 40)
print(df_retail_2.isna().sum())
print('*' * 120)


************************************************************************************************************************
****************************************df_retail------- revisión de valores ausentes****************************************
num_factura             0
cod_articulo            0
descripcion          1454
cantidad                0
fh_factura              0
precio_initario         0
id_cliente         135037
pais                    0
dtype: int64
************************************************************************************************************************


In [22]:
print('*' * 120)
print(df_retail_2.info())
print('*' * 120)

************************************************************************************************************************
<class 'pandas.DataFrame'>
Index: 536641 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column           Non-Null Count   Dtype         
---  ------           --------------   -----         
 0   num_factura      536641 non-null  object        
 1   cod_articulo     536641 non-null  object        
 2   descripcion      535187 non-null  object        
 3   cantidad         536641 non-null  int64         
 4   fh_factura       536641 non-null  datetime64[us]
 5   precio_unitario  536641 non-null  float64       
 6   id_cliente       401604 non-null  Int64         
 7   pais             536641 non-null  str           
dtypes: Int64(1), datetime64[us](1), float64(1), int64(1), object(3), str(1)
memory usage: 37.4+ MB
None
************************************************************************************************************************


In [23]:
# Calcular el número de valores ausentes
print(df_retail_2.isna().sum())

num_factura             0
cod_articulo            0
descripcion          1454
cantidad                0
fh_factura              0
precio_unitario         0
id_cliente         135037
pais                    0
dtype: int64


### Si quiero ver las filas que tienen valores ausentes

In [25]:
df_retail_2[df_retail_2['descripcion'].isna()].head(10)

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_unitario,id_cliente,pais
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.0,<NA>,United Kingdom
1970,536545,21134,NaN,1,2010-12-01 14:32:00,0.0,<NA>,United Kingdom
1971,536546,22145,NaN,1,2010-12-01 14:33:00,0.0,<NA>,United Kingdom
1972,536547,37509,NaN,1,2010-12-01 14:33:00,0.0,<NA>,United Kingdom
1987,536549,85226A,NaN,1,2010-12-01 14:34:00,0.0,<NA>,United Kingdom
1988,536550,85044,NaN,1,2010-12-01 14:34:00,0.0,<NA>,United Kingdom
2024,536552,20950,NaN,1,2010-12-01 14:34:00,0.0,<NA>,United Kingdom
2025,536553,37461,NaN,3,2010-12-01 14:35:00,0.0,<NA>,United Kingdom
2026,536554,84670,NaN,23,2010-12-01 14:35:00,0.0,<NA>,United Kingdom
2406,536589,21777,NaN,-10,2010-12-01 16:50:00,0.0,<NA>,United Kingdom


Aqui hagamos una pausa y recordemos lo de antes se menciono 🤔🤔🤔🤔🤔🤔🤔🤔🤔🤔🤔🤔🤔🤔🤔🤔🤔🤔🤔🤔

* Una cantidad negativa podría ser una devolución.

* Un precio cero podría corresponder a una operación especial.

* Un cliente ausente no significa necesariamente que la venta sea inválida.

* Una descripción ausente sí es un problema de calidad, pero su impacto depende del análisis que quieras realizar.

vamos a ver como ir obteniendo la descripcion por la del cod_articulo


In [ ]:
#antes de ver que me este faltando preguntemos por este registro solamente: 22139

In [33]:
descripciones_por_codigo = (
    df_retail_2[df_retail_2['descripcion'].notna()]
    .groupby('cod_articulo')['descripcion']
    .nunique()
)

codigos_con_varias_descripciones = descripciones_por_codigo[
    descripciones_por_codigo > 1
]

print(codigos_con_varias_descripciones)

cod_articulo
10080           2
10133           2
16008           2
16045           2
20622           2
               ..
90195A          2
90210D          2
DCGS0003        2
DCGS0069        2
gift_0001_20    2
Name: descripcion, Length: 650, dtype: int64


In [34]:
# Crear un diccionario con el código y su descripción
mapa_descripciones = (
    df_retail_2.dropna(subset=['descripcion'])
    .drop_duplicates(subset=['cod_articulo'])
    .set_index('cod_articulo')['descripcion']
)

# Completar únicamente las descripciones ausentes
df_retail_2['descripcion'] = (
    df_retail_2['descripcion'].fillna(
        df_retail_2['cod_articulo'].map(mapa_descripciones)
    )
)

In [35]:
ausentes_antes = df_retail_2['descripcion'].isna().sum()

print("Descripciones ausentes:", ausentes_antes)

Descripciones ausentes: 112


In [36]:
print(
    "Descripciones ausentes después:",
    df_retail_2['descripcion'].isna().sum()
)

Descripciones ausentes después: 112


In [37]:
# Calcular el número de valores ausentes
print(df_retail_2.isna().sum())

num_factura             0
cod_articulo            0
descripcion           112
cantidad                0
fh_factura              0
precio_unitario         0
id_cliente         135037
pais                    0
dtype: int64


In [31]:
print(df_retail_2.isna().sum())

num_factura             0
cod_articulo            0
descripcion          1454
cantidad                0
fh_factura              0
precio_unitario         0
id_cliente         135037
pais                    0
dtype: int64


Aqui se esta observando algo que el precio unitario esta en 0.0, por lo que procedemos agregarle un filtro de que sea mayor a 0.0


,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_initario,id_cliente,pais


Despues de esto observamos que no hay es necesario estos valores por lo que aunque dicen que cantidad algunos son mayores a 1 unidad por lo cual no se verea una afectacion si quitamos completamente estos registro


In [25]:
df_retail[df_retail['precio_initario'].isna()].head(10)

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_initario,id_cliente,pais


In [26]:
# hagamos un respaldo antes 
df_retail_2 = df_retail

#ahora si eliminemos esos registros que se ve que no afecta que podamos eliminarlos
df_retail_2[df_retail_2['descripcion'].isna()].head(10)

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_initario,id_cliente,pais
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.0,<NA>,United Kingdom
1970,536545,21134,NaN,1,2010-12-01 14:32:00,0.0,<NA>,United Kingdom
1971,536546,22145,NaN,1,2010-12-01 14:33:00,0.0,<NA>,United Kingdom
1972,536547,37509,NaN,1,2010-12-01 14:33:00,0.0,<NA>,United Kingdom
1987,536549,85226A,NaN,1,2010-12-01 14:34:00,0.0,<NA>,United Kingdom
1988,536550,85044,NaN,1,2010-12-01 14:34:00,0.0,<NA>,United Kingdom
2024,536552,20950,NaN,1,2010-12-01 14:34:00,0.0,<NA>,United Kingdom
2025,536553,37461,NaN,3,2010-12-01 14:35:00,0.0,<NA>,United Kingdom
2026,536554,84670,NaN,23,2010-12-01 14:35:00,0.0,<NA>,United Kingdom
2406,536589,21777,NaN,-10,2010-12-01 16:50:00,0.0,<NA>,United Kingdom


Eliminemos los duplicados explicitamente

In [27]:
df_retail_2.drop_duplicates(inplace=True)

In [32]:
print(df_retail_2.duplicated().sum())

0


Eliminemos los valores ausentes, revisemos antes si esos asusenten cumplen que sea mayor de 0.0 en precio unitario

In [35]:
df_retail_2[df_retail_2['descripcion'].isna() &
          df_retail_2['precio_initario']>0.0].head(10)

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_initario,id_cliente,pais


In [37]:
df_retail_2 = df_retail_2.drop(
    df_retail_2[
        df_retail_2['descripcion'].isna() &
        (df_retail_2['precio_initario'] == 0.0)
    ].index
)

In [38]:
# Calcular el número de valores ausentes
print(df_retail_2.isna().sum())

num_factura             0
cod_articulo            0
descripcion             0
cantidad                0
fh_factura              0
precio_initario         0
id_cliente         133583
pais                    0
dtype: int64


In [48]:
print(df_retail_2.head())

  num_factura cod_articulo                          descripcion  cantidad  \
0      536365       85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1      536365        71053                  WHITE METAL LANTERN         6   
2      536365       84406B       CREAM CUPID HEARTS COAT HANGER         8   
3      536365       84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4      536365       84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           fh_factura  precio_initario id_cliente            pais  
0 2010-12-01 08:26:00             2.55    17850.0  United Kingdom  
1 2010-12-01 08:26:00             3.39    17850.0  United Kingdom  
2 2010-12-01 08:26:00             2.75    17850.0  United Kingdom  
3 2010-12-01 08:26:00             3.39    17850.0  United Kingdom  
4 2010-12-01 08:26:00             3.39    17850.0  United Kingdom  


In [56]:
df_retail_2[df_retail_2['id_cliente'].isna() &
          (df_retail_2['precio_initario']==0)].head(10)

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_initario,id_cliente,pais
6391,536941,22734,amazon,20,2010-12-03 12:08:00,0.0,<NA>,United Kingdom
6392,536942,22139,amazon,15,2010-12-03 12:08:00,0.0,<NA>,United Kingdom
7313,537032,21275,?,-30,2010-12-03 16:50:00,0.0,<NA>,United Kingdom
13217,537425,84968F,check,-20,2010-12-06 15:35:00,0.0,<NA>,United Kingdom
13218,537426,84968E,check,-35,2010-12-06 15:36:00,0.0,<NA>,United Kingdom
13264,537432,35833G,damages,-43,2010-12-06 16:10:00,0.0,<NA>,United Kingdom
14335,537534,85064,CREAM SWEETHEART LETTER RACK,1,2010-12-07 11:48:00,0.0,<NA>,United Kingdom
14336,537534,84832,ZINC WILLIE WINKIE CANDLE STICK,1,2010-12-07 11:48:00,0.0,<NA>,United Kingdom
14337,537534,84692,BOX OF 24 COCKTAIL PARASOLS,2,2010-12-07 11:48:00,0.0,<NA>,United Kingdom
14338,537534,48184,DOORMAT ENGLISH ROSE,3,2010-12-07 11:48:00,0.0,<NA>,United Kingdom


In [58]:
df_retail_2[df_retail_2['id_cliente'].isna() &
            (df_retail_2['cantidad']<0) &
          (df_retail_2['precio_initario']==0)].head(10)

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_initario,id_cliente,pais
7313,537032,21275,?,-30,2010-12-03 16:50:00,0.0,<NA>,United Kingdom
13217,537425,84968F,check,-20,2010-12-06 15:35:00,0.0,<NA>,United Kingdom
13218,537426,84968E,check,-35,2010-12-06 15:36:00,0.0,<NA>,United Kingdom
13264,537432,35833G,damages,-43,2010-12-06 16:10:00,0.0,<NA>,United Kingdom
21338,538072,22423,faulty,-13,2010-12-09 14:10:00,0.0,<NA>,United Kingdom
21518,538090,20956,?,-723,2010-12-09 14:48:00,0.0,<NA>,United Kingdom
22296,538161,46000S,Dotcom sales,-100,2010-12-09 17:25:00,0.0,<NA>,United Kingdom
22297,538162,46000M,Dotcom sales,-100,2010-12-09 17:25:00,0.0,<NA>,United Kingdom
42564,540010,22501,reverse 21/5/10 adjustment,-100,2011-01-04 11:13:00,0.0,<NA>,United Kingdom
42566,540012,22502,reverse 21/5/10 adjustment,-100,2011-01-04 11:14:00,0.0,<NA>,United Kingdom


In [63]:
# se observan varios registros, pero tambien quisiera saber cuantos son los registro que hay en total y para eso realizo esto

total =df_retail_2[
    df_retail_2['id_cliente'].isna() &
    (df_retail_2['cantidad']<0) &
    (df_retail_2['precio_initario']==0)]

print("cantidad que cumple esta condicion es :" , total.shape[0])

cantidad que cumple esta condicion es : 0


Viendolo de esa manera combiene mas eliminarlo y ver que queda

In [62]:
df_retail_2 = df_retail_2.drop(
    df_retail_2[
        df_retail_2['id_cliente'].isna() &
    (df_retail_2['cantidad']<0) &
    (df_retail_2['precio_initario']==0)
    ].index
)

In [64]:
total =df_retail_2[
    df_retail_2['id_cliente'].isna() &
    (df_retail_2['cantidad']<0) &
    (df_retail_2['precio_initario']==0)]

print("cantidad que cumple esta condicion es :" , total.shape[0])

cantidad que cumple esta condicion es : 0


In [65]:
# Calcular el número de valores ausentes
print(df_retail_2.isna().sum())

num_factura             0
cod_articulo            0
descripcion             0
cantidad                0
fh_factura              0
precio_initario         0
id_cliente         133109
pais                    0
dtype: int64


In [67]:
df_retail_2[df_retail_2['id_cliente'].isna() &
            
          (df_retail_2['precio_initario']==0)].head(10)

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_initario,id_cliente,pais
6391,536941,22734,amazon,20,2010-12-03 12:08:00,0.0,<NA>,United Kingdom
6392,536942,22139,amazon,15,2010-12-03 12:08:00,0.0,<NA>,United Kingdom
14335,537534,85064,CREAM SWEETHEART LETTER RACK,1,2010-12-07 11:48:00,0.0,<NA>,United Kingdom
14336,537534,84832,ZINC WILLIE WINKIE CANDLE STICK,1,2010-12-07 11:48:00,0.0,<NA>,United Kingdom
14337,537534,84692,BOX OF 24 COCKTAIL PARASOLS,2,2010-12-07 11:48:00,0.0,<NA>,United Kingdom
14338,537534,48184,DOORMAT ENGLISH ROSE,3,2010-12-07 11:48:00,0.0,<NA>,United Kingdom
14339,537534,48111,DOORMAT 3 SMILEY CATS,1,2010-12-07 11:48:00,0.0,<NA>,United Kingdom
14340,537534,22697,GREEN REGENCY TEACUP AND SAUCER,1,2010-12-07 11:48:00,0.0,<NA>,United Kingdom
14341,537534,22682,FRENCH BLUE METAL DOOR SIGN 7,1,2010-12-07 11:48:00,0.0,<NA>,United Kingdom
14342,537534,22680,FRENCH BLUE METAL DOOR SIGN 5,1,2010-12-07 11:48:00,0.0,<NA>,United Kingdom


In [69]:
total =df_retail_2[
    df_retail_2['id_cliente'].isna() &
  
    (df_retail_2['precio_initario']==0)]

print("cantidad que cumple esta condicion es :" , total.shape[0])

cantidad que cumple esta condicion es : 542


In [71]:
df_retail_2 = df_retail_2.drop(
    df_retail_2[
    df_retail_2['id_cliente'].isna() &
    (df_retail_2['precio_initario']==0)
    ].index
)

In [73]:
#volvemos a revisar cuantos valores ausentes hay, antes de quitar esos registros habia 542, veamos ahora cuantos quedan
# Calcular el número de valores ausentes
print(df_retail_2.isna().sum())

num_factura             0
cod_articulo            0
descripcion             0
cantidad                0
fh_factura              0
precio_initario         0
id_cliente         132567
pais                    0
dtype: int64


In [77]:
df_retail_2[df_retail_2['id_cliente'].isna()].head(10)

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_initario,id_cliente,pais
1443,536544,21773,DECORATIVE ROSE BATHROOM BOTTLE,1,2010-12-01 14:32:00,2.51,<NA>,United Kingdom
1444,536544,21774,DECORATIVE CATS BATHROOM BOTTLE,2,2010-12-01 14:32:00,2.51,<NA>,United Kingdom
1445,536544,21786,POLKADOT RAIN HAT,4,2010-12-01 14:32:00,0.85,<NA>,United Kingdom
1446,536544,21787,RAIN PONCHO RETROSPOT,2,2010-12-01 14:32:00,1.66,<NA>,United Kingdom
1447,536544,21790,VINTAGE SNAP CARDS,9,2010-12-01 14:32:00,1.66,<NA>,United Kingdom
1448,536544,21791,VINTAGE HEADS AND TAILS CARD GAME,2,2010-12-01 14:32:00,2.51,<NA>,United Kingdom
1449,536544,21801,CHRISTMAS TREE DECORATION WITH BELL,10,2010-12-01 14:32:00,0.43,<NA>,United Kingdom
1450,536544,21802,CHRISTMAS TREE HEART DECORATION,9,2010-12-01 14:32:00,0.43,<NA>,United Kingdom
1451,536544,21803,CHRISTMAS TREE STAR DECORATION,11,2010-12-01 14:32:00,0.43,<NA>,United Kingdom
1452,536544,21809,CHRISTMAS HANGING TREE WITH BELL,1,2010-12-01 14:32:00,2.51,<NA>,United Kingdom


In [72]:
df_retail_2[
    df_retail_2['id_cliente'].isna() &
  
    (df_retail_2['precio_initario']==0)]

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_initario,id_cliente,pais


In [88]:
total =df_retail_2[
    df_retail_2['id_cliente'].isna() &
  
    (df_retail_2['cantidad']==-1)
    ].sort_values(
    by='precio_initario',
    ascending=False)

#print("cantidad que cumple esta condicion es :" , total.shape[0])
#print("cantidad que cumple esta condicion es :" , total)
print(total.head(10))
#print(total.sum())

print("cantidad que cumple esta condicion es :" , total.shape[0])


Empty DataFrame
Columns: [num_factura, cod_articulo, descripcion, cantidad, fh_factura, precio_initario, id_cliente, pais]
Index: []
cantidad que cumple esta condicion es : 0


In [ ]:
#se eliminaron los registros donde la cantidad es -1 por que no tiene sentido dejar este tipo de registro por mucho que tenga un gran numero para el precio unitario

df_retail_2 = df_retail_2.drop(
    df_retail_2[
    df_retail_2['id_cliente'].isna() &
    (df_retail_2['cantidad']==-1)
    ].index
)

In [89]:
print(df_retail_2.isna().sum())

num_factura             0
cod_articulo            0
descripcion             0
cantidad                0
fh_factura              0
precio_initario         0
id_cliente         132311
pais                    0
dtype: int64


### Detangamos todo por que podemos dar otro enfoque

In [92]:
print("DataFrame original:", df_retail.shape)
print("DataFrame de trabajo:", df_retail_2.shape)

DataFrame original: (536641, 8)
DataFrame de trabajo: (533915, 8)


In [91]:
df_retail_2[df_retail_2['id_cliente'].isna() &
          df_retail_2['precio_initario']!=0].head(10)

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_initario,id_cliente,pais
1443,536544,21773,DECORATIVE ROSE BATHROOM BOTTLE,1,2010-12-01 14:32:00,2.51,<NA>,United Kingdom
1444,536544,21774,DECORATIVE CATS BATHROOM BOTTLE,2,2010-12-01 14:32:00,2.51,<NA>,United Kingdom
1445,536544,21786,POLKADOT RAIN HAT,4,2010-12-01 14:32:00,0.85,<NA>,United Kingdom
1446,536544,21787,RAIN PONCHO RETROSPOT,2,2010-12-01 14:32:00,1.66,<NA>,United Kingdom
1447,536544,21790,VINTAGE SNAP CARDS,9,2010-12-01 14:32:00,1.66,<NA>,United Kingdom
1448,536544,21791,VINTAGE HEADS AND TAILS CARD GAME,2,2010-12-01 14:32:00,2.51,<NA>,United Kingdom
1449,536544,21801,CHRISTMAS TREE DECORATION WITH BELL,10,2010-12-01 14:32:00,0.43,<NA>,United Kingdom
1450,536544,21802,CHRISTMAS TREE HEART DECORATION,9,2010-12-01 14:32:00,0.43,<NA>,United Kingdom
1451,536544,21803,CHRISTMAS TREE STAR DECORATION,11,2010-12-01 14:32:00,0.43,<NA>,United Kingdom
1452,536544,21809,CHRISTMAS HANGING TREE WITH BELL,1,2010-12-01 14:32:00,2.51,<NA>,United Kingdom


In [45]:
df_retail_2[df_retail_2['id_cliente'].isna()].head(10)

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_initario,id_cliente,pais
1443,536544,21773,DECORATIVE ROSE BATHROOM BOTTLE,1,2010-12-01 14:32:00,2.51,<NA>,United Kingdom
1444,536544,21774,DECORATIVE CATS BATHROOM BOTTLE,2,2010-12-01 14:32:00,2.51,<NA>,United Kingdom
1445,536544,21786,POLKADOT RAIN HAT,4,2010-12-01 14:32:00,0.85,<NA>,United Kingdom
1446,536544,21787,RAIN PONCHO RETROSPOT,2,2010-12-01 14:32:00,1.66,<NA>,United Kingdom
1447,536544,21790,VINTAGE SNAP CARDS,9,2010-12-01 14:32:00,1.66,<NA>,United Kingdom
1448,536544,21791,VINTAGE HEADS AND TAILS CARD GAME,2,2010-12-01 14:32:00,2.51,<NA>,United Kingdom
1449,536544,21801,CHRISTMAS TREE DECORATION WITH BELL,10,2010-12-01 14:32:00,0.43,<NA>,United Kingdom
1450,536544,21802,CHRISTMAS TREE HEART DECORATION,9,2010-12-01 14:32:00,0.43,<NA>,United Kingdom
1451,536544,21803,CHRISTMAS TREE STAR DECORATION,11,2010-12-01 14:32:00,0.43,<NA>,United Kingdom
1452,536544,21809,CHRISTMAS HANGING TREE WITH BELL,1,2010-12-01 14:32:00,2.51,<NA>,United Kingdom


In [39]:
#ahora si eliminemos esos registros que se ve que no afecta que podamos eliminarlos
df_retail_2[df_retail_2['id_cliente'].isna()].head(10)

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_initario,id_cliente,pais
1443,536544,21773,DECORATIVE ROSE BATHROOM BOTTLE,1,2010-12-01 14:32:00,2.51,<NA>,United Kingdom
1444,536544,21774,DECORATIVE CATS BATHROOM BOTTLE,2,2010-12-01 14:32:00,2.51,<NA>,United Kingdom
1445,536544,21786,POLKADOT RAIN HAT,4,2010-12-01 14:32:00,0.85,<NA>,United Kingdom
1446,536544,21787,RAIN PONCHO RETROSPOT,2,2010-12-01 14:32:00,1.66,<NA>,United Kingdom
1447,536544,21790,VINTAGE SNAP CARDS,9,2010-12-01 14:32:00,1.66,<NA>,United Kingdom
1448,536544,21791,VINTAGE HEADS AND TAILS CARD GAME,2,2010-12-01 14:32:00,2.51,<NA>,United Kingdom
1449,536544,21801,CHRISTMAS TREE DECORATION WITH BELL,10,2010-12-01 14:32:00,0.43,<NA>,United Kingdom
1450,536544,21802,CHRISTMAS TREE HEART DECORATION,9,2010-12-01 14:32:00,0.43,<NA>,United Kingdom
1451,536544,21803,CHRISTMAS TREE STAR DECORATION,11,2010-12-01 14:32:00,0.43,<NA>,United Kingdom
1452,536544,21809,CHRISTMAS HANGING TREE WITH BELL,1,2010-12-01 14:32:00,2.51,<NA>,United Kingdom


In [49]:
# Calcular el número de valores ausentes
print(df_retail_2.isna().sum())

num_factura             0
cod_articulo            0
descripcion             0
cantidad                0
fh_factura              0
precio_initario         0
id_cliente         133583
pais                    0
dtype: int64


In [33]:
df_retail_2[df_retail_2['descripcion'].isna()].head(10)

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_initario,id_cliente,pais
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.0,<NA>,United Kingdom
1970,536545,21134,NaN,1,2010-12-01 14:32:00,0.0,<NA>,United Kingdom
1971,536546,22145,NaN,1,2010-12-01 14:33:00,0.0,<NA>,United Kingdom
1972,536547,37509,NaN,1,2010-12-01 14:33:00,0.0,<NA>,United Kingdom
1987,536549,85226A,NaN,1,2010-12-01 14:34:00,0.0,<NA>,United Kingdom
1988,536550,85044,NaN,1,2010-12-01 14:34:00,0.0,<NA>,United Kingdom
2024,536552,20950,NaN,1,2010-12-01 14:34:00,0.0,<NA>,United Kingdom
2025,536553,37461,NaN,3,2010-12-01 14:35:00,0.0,<NA>,United Kingdom
2026,536554,84670,NaN,23,2010-12-01 14:35:00,0.0,<NA>,United Kingdom
2406,536589,21777,NaN,-10,2010-12-01 16:50:00,0.0,<NA>,United Kingdom


In [30]:
print(df_retail_2.isna().sum())

num_factura             0
cod_articulo            0
descripcion          1454
cantidad                0
fh_factura              0
precio_initario         0
id_cliente         135037
pais                    0
dtype: int64


In [ ]:
# hagamos un respaldo antes 
df_retail_2 = df_retail

#ahora si eliminemos esos registros que se ve que no afecta que podamos eliminarlos
df_retail_2[df_retail_2['descripcion'].isna()].head(10)

In [ ]:

print('*' * 80)
print('cuenta con valores null--->', df_retail.isnull().sum())
print('*' * 80)
print('*' * 80)

Etapa 2. Preprocesamiento de datos

In [ ]:





print('*' * 80)
print('cuenta con valores null--->', df_retail.isnull().sum())
print('*' * 80)
print('*' * 80)



df_retail['id_cliente'] = df_retail['id_cliente'].astype('string')
df_retail['pais'] = df_retail['pais'].astype('string')



print('*' * 80)
print(df_retail.info())
print('*' * 80)
print('*' * 40 +'df_retail------- revisión de valores ausentes' + '*' * 40)
print(df_retail.isna().sum())




'''

# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'
#columnas a tratar
colum_cambiar =['track', 'artist','genre']
for col in df.columns:
    if col in colum_cambiar:
        df[col]=df[col].fillna('unknown', inplace =False)
print(df.isna().sum())


'''

ausentes = pd.DataFrame({
    'cantidad': df_retail.isna().sum(),
    'porcentaje': (df_retail.isna().sum() / len(df_retail) * 100).round(2),
    'porcentaje %': (((df_retail.isna().sum() / len(df_retail)) * 100).round(2))*100

})

print(ausentes)

#se observa que hay muy un numero muy mayor de id de cliente ausentes, por lo que haremos procedemos a quitarlos
#print(df_retail.isna.head())
print(df_retail[df_retail['id_cliente'].isna()].head(10))

#para saber cuantos son los registros con <NA>
print(df_retail['id_cliente'].isna().sum())

# veamos los id_cliente de que pasi son :

print(
    df_retail[df_retail['id_cliente'].isna()]['pais'].value_counts()
)
'''

# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'
#columnas a tratar
colum_cambiar =['descripcion', 'id_cliente']
for col in df_retail.columns:
    if col in colum_cambiar:
        df_retail[col]=df_retail[col].fillna('unknown', inplace =False)
print(df_retail.isna().sum())



df_retail


'''

********************************************************************************
****************************************df_retail------- solo los 5 registros****************************************
  num_factura cod_articulo                          descripcion  cantidad  \
0      536365       85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1      536365        71053                  WHITE METAL LANTERN         6   
2      536365       84406B       CREAM CUPID HEARTS COAT HANGER         8   
3      536365       84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4      536365       84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           fh_factura  precio_initario  id_cliente            pais  
0 2010-12-01 08:26:00             2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00             3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00             2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00             3.39     17850.0  United Kin

"\n\n# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'\n#columnas a tratar\ncolum_cambiar =['descripcion', 'id_cliente']\nfor col in df_retail.columns:\n    if col in colum_cambiar:\n        df_retail[col]=df_retail[col].fillna('unknown', inplace =False)\nprint(df_retail.isna().sum())\n\n\n\ndf_retail\n\n\n"